src/prefect/futures.py

```diff
@@ -171,43 +177,80 @@ def task_run(self):
     def task_run(self, task_run):
         self._task_run = task_run
 
-    def wait(
-        self, timeout: Optional[float] = None, polling_interval: Optional[float] = 0.2
-    ) -> None:
-        start_time = time.time()
-        # TODO: Websocket implementation?
-        while True:
-            self.task_run = cast(
-                TaskRun, self.client.read_task_run(task_run_id=self.task_run_id)
+    def wait(self, timeout: Optional[float] = None) -> None:
+        return run_coro_as_sync(self.wait_async(timeout=timeout))
+
+    async def wait_async(self, timeout: Optional[float] = None):
+        if self._final_state:
+            logger.debug(
+                "Final state already set for %s. Returning...", self.task_run.task_key
             )
-            if self.task_run.state and self.task_run.state.is_final():
+            return
+
+        # Read task run to see if it is still running
+        async with get_client() as client:
+            self.task_run = await client.read_task_run(task_run_id=self._task_run_id)
+            if self.task_run.state.is_final():
+                logger.debug(
+                    "Task run %s already finished. Returning...", self.task_run.task_key
+                )
                 self._final_state = self.task_run.state
                 return
-            if timeout is not None and (time.time() - start_time) > timeout:
-                return
-            time.sleep(polling_interval)
+
+            # If still running, wait for a completed event from the server
+            with anyio.move_on_after(delay=timeout):
+                async with get_events_subscriber(
+                    filter=EventFilter(
+                        event=EventNameFilter(
+                            name=[
+                                f"prefect.task-run.{state.name.title()}"
+                                for state in TERMINAL_STATES
+                            ],
+                        ),
+                        resource=EventResourceFilter(
+                            id=[f"prefect.task-run.{self.task_run_id}"]
+                        ),
+                    )
+                ) as subscriber:
+                    logger.debug(
+                        "Waiting for completed event for %s...", self.task_run.task_key
+                    )
+                    async for event in subscriber:
+                        self.task_run = await client.read_task_run(
```

Comment:
I wonder if it could happen in Cloud that we receive the event and then go to read the task run and get a stale version from a read replica. Is there any way the event can send the state in its payload so that we could see that?

Reply:
Yeah the event includes the state, but I can also include a check for `.is_final` when reading the final state. I'm not sure what we can do if we get stale data, though.

Reply:
I would use the state from the event then, if we know that one's the true final state but there's this chance of getting a stale running state back. Because when we see the completed event we know that's the final completed state. (It sounds like)

Reply:
Also, I mean, for purely performance reasons, if we have the state we don't need to read it from the API I guess

Reply:
Ohhhh right, this future has the entire task run object

Reply:
I will check to see if we get enough info from the event to construct a full state object. I also am going to remote the `task_run` attribute from this future because I don't think it's strictly necessary.